Libs/vtkAddon/vtkStreamingVolumeCodecFactory.h

```diff
@@ -0,0 +1,116 @@
+/*=auto=========================================================================
+
+  Portions (c) Copyright 2005 Brigham and Women's Hospital (BWH) All Rights Reserved.
+
+  See COPYRIGHT.txt
+  or http://www.slicer.org/copyright/copyright.txt for details.
+
+  Program:   3D Slicer
+  Module:    $RCSfile: vtkStreamingVolumeCodecFactory.h,v $
+  Date:      $Date: 2006/03/19 17:12:29 $
+  Version:   $Revision: 1.13 $
+
+=========================================================================auto=*/
+
+
+#ifndef __vtkStreamingVolumeCodecFactory_h
+#define __vtkStreamingVolumeCodecFactory_h
+
+// VTK includes
+#include <vtkObject.h>
+#include <vtkSmartPointer.h>
+
+// STD includes
+#include <map>
+
+// vtkAddon includes
+#include "vtkAddon.h"
+#include "vtkStreamingVolumeCodec.h"
+
+/// \ingroup Volumes
+/// \brief Class that can create compresion device for streaming volume instances.
+///
+/// This singleton class is a repository of all compression codecs for compressing volume .
+/// Singleton pattern adopted from vtkEventBroker class.
+class VTK_ADDON_EXPORT vtkStreamingVolumeCodecFactory : public vtkObject
+{
+public:
+
+  vtkTypeMacro(vtkStreamingVolumeCodecFactory, vtkObject);
+  void PrintSelf(ostream& os, vtkIndent indent) VTK_OVERRIDE;
+
+  /// Registers a new video compression codec
+  /// \param newCodecPointer Function pointer to the codec type new function (e.g. (PointerToCodecBaseNew)&vtkStreamingVolumeCodec::New )
+  bool RegisterStreamingCodec(vtkSmartPointer<vtkStreamingVolumeCodec> codec);
+
+  /// Removes a codec from the factory
+  /// This does not affect codecs that have already been instantiated
+  /// \param codecClassName full name of the codec class that is being unregistered
+  /// Returns true if the codec is successfully unregistered
+  bool UnregisterStreamingCodecByClassName(const std::string& codecClassName);
+
+  /// Get pointer to codec new function, or NULL if the codec type not registered
+  /// Usage: vtkSmartPointer<vtkStreamingVolumeCodec> codec = GetVolumeCodecNewPointerByType("igtlioVideoDevice")();
+  /// Returns NULL if no matching codec can be found
+  vtkStreamingVolumeCodec* CreateCodecByClassName(const std::string& codecClassName);
+
+  /// Creates a new codec that matches the specified FourCC code
+  /// \param codecFourCC FourCC representing the encoding method
+  /// See https://www.fourcc.org/codecs.php for an incomplete list
+  /// Returns NULL if no matching codec can be found
+  vtkStreamingVolumeCodec* CreateCodecByFourCC(const std::string codecFourCC);
+
+  /// Returns a list of all registered Codecs
+  const std::vector<vtkSmartPointer<vtkStreamingVolumeCodec> >& GetStreamingCodecClasses();
```

Comment:
It would be a bit nicer and safer if we returned the class names instead of classes themselves